File: AGTIA_Objective-C/QQCuteView/QQCuteView/Classes/QQCuteView.m
//
//  QQCuteView.m
//  QQCuteView
//
//  Created by fernando on 2017/2/6.
//  Copyright © 2017年 fernando. All rights reserved.
//

#import "QQCuteView.h"

@implementation QQCuteView {
    UIBezierPath *cutePath;
    UIColor *fillColorForCute; // 气泡填充颜色
    UIDynamicAnimator *animator;
    UISnapBehavior  *snap; // dynamic 震荡效果 UISnapBehavior 震荡行为 荡行为定义了一个动态运动，达到指定点后实现震荡效果，震荡的数量是可以设置的。 配置属性： damping － 动画结束的时候的震荡值 0.0-1.0；其中0.0最小，1.0最大。默认0.5
    
    UIView *backView;
    CGFloat r1; // 小圆半径
    CGFloat r2; // 大圆半径
    CGFloat x1; // 小圆中心x坐标
    CGFloat y1; // 小圆中心y坐标
    CGFloat x2; // 大圆中心x坐标
    CGFloat y2; // 大圆中心y坐标
    CGFloat centerDistance; // 俩圆心之间的距离
    CGFloat cosDigree;
    CGFloat sinDigree;
    
    // http://kittenyang.com/content/images/2015/Mar/Screen-Shot-2015-03-03-at-14-25-22.png
    CGPoint pointA; //A
    CGPoint pointB; //B
    CGPoint pointD; //D
    CGPoint pointC; //C
    CGPoint pointO; //O
    CGPoint pointP; //P
    
    CGRect oldBackViewFrame;
    CGPoint initialPoint;
    CGPoint oldBackViewCenter;
    CAShapeLayer *shapeLayer;
}

- (id)initWithPoint:(CGPoint)point superView:(UIView *)sView {
    self = [super initWithFrame:CGRectMake(point.x, point.y, self.bubbleWidth, self.bubbleWidth)];
    if (self) {
        initialPoint = point;
        self.containerView = sView;
        [self.containerView addSubview:self];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setUp];
    }
    return self;
}

- (void)drawRect {
    x1 = backView.center.x;
    y1 = backView.center.y;
    x2 = self.frontView.center.x;
    y2 = self.frontView.center.y;
    
    centerDistance = sqrtf((x2-x1)*(x2-x1) + (y2-y1)*(y2-y1));
    
    if (centerDistance == 0) {
        cosDigree = 1;
        sinDigree = 0;
    } else {
        cosDigree = (y2 - y1)/centerDistance;
        sinDigree = (x2 - x1)/centerDistance;
    }
    
    r1 = oldBackViewFrame.size.width/2 - centerDistance/self.viscosity;
    
    pointA = CGPointMake(x1 - r1 * cosDigree, y1 + r1 * sinDigree);
    pointB = CGPointMake(x1 + r1 * cosDigree, y1 - r1 * sinDigree);
    pointD = CGPointMake(x2 - r2 * cosDigree, y2 + r2 * sinDigree);
    pointC = CGPointMake(x2 + r2 * cosDigree, y2 - r2 * sinDigree);
    
    pointO = CGPointMake(pointA.x + (centerDistance / 2)*sinDigree, pointA.y + (centerDistance / 2)*cosDigree);
    pointP = CGPointMake(pointB.x + (centerDistance / 2)*sinDigree, pointB.y + (centerDistance / 2)*cosDigree);
    
    backView.center = oldBackViewCenter;
    backView.bounds = CGRectMake(0, 0, r1*2, r2*2);
    backView.layer.cornerRadius = r1;
    
    cutePath = [UIBezierPath bezierPath];
    [cutePath moveToPoint:pointA];
    [cutePath addQuadCurveToPoint:pointD controlPoint:pointO];
    [cutePath moveToPoint:pointC];
    [cutePath addQuadCurveToPoint:pointB controlPoint:pointP];
    [cutePath moveToPoint:pointA];

    if (!backView.hidden) {
        shapeLayer.path = cutePath.CGPath;
        shapeLayer.fillColor = fillColorForCute.CGColor;
        [self.containerView.layer insertSublayer:shapeLayer below:self.frontView.layer];
    }
}

- (void)setUp {
    shapeLayer = [CAShapeLayer layer];
    self.backgroundColor = [UIColor clearColor];
    self.frontView = [[UIView alloc] initWithFrame:CGRectMake(initialPoint.x, initialPoint.y, self.bubbleWidth, self.bubbleWidth)];
    
    r2 = self.frontView.bounds.size.width/2;
    self.frontView.layer.cornerRadius = r2;
    self.frontView.backgroundColor = self.bubbleColor;

    backView = [[UIView alloc] initWithFrame:self.frontView.frame];
    r1 = backView.bounds.size.width/2;
    backView.layer.cornerRadius = r1;
    backView.backgroundColor = self.bubbleColor;
    
    self.bubbleLabel = [[UILabel alloc]init];
    self.bubbleLabel.frame = CGRectMake(0, 0, self.frontView.bounds.size.width, self.frontView.bounds.size.height);
    self.bubbleLabel.textColor = [UIColor whiteColor];
    self.bubbleLabel.textAlignment = NSTextAlignmentCenter;

    [self.frontView insertSubview:self.bubbleLabel atIndex:0];

    [self.containerView addSubview:backView];
    [self.containerView addSubview:self.frontView];

    x1 = backView.center.x;
    y1 = backView.center.y;
    x2 = self.frontView.center.x;
    y2 = self.frontView.center.y;

    pointA = CGPointMake(x1-r1,y1);   // A
    pointB = CGPointMake(x1+r1, y1);  // B
    pointD = CGPointMake(x2-r2, y2);  // D
    pointC = CGPointMake(x2+r2, y2);  // C
    pointO = CGPointMake(x1-r1,y1);   // O
    pointP = CGPointMake(x2+r2, y2);  // P
//
    oldBackViewFrame = backView.frame;
    oldBackViewCenter = backView.center;
//
    backView.hidden = YES;//为了看到 frontView 的气泡晃动效果，需要暂时隐藏 backView
    [self addAniamtionLikeGameCenterBubble];
//
    UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc]initWithTarget:self action:@selector(handleDragGesture:)];
    [self.frontView addGestureRecognizer:pan];
}

- (void)handleDragGesture:(UIPanGestureRecognizer *)ges {
    CGPoint dragPoint = [ges locationInView:self.containerView];
    if (ges.state == UIGestureRecognizerStateBegan) {
        backView.hidden = NO;
        fillColorForCute = self.bubbleColor;
        [self RemoveAniamtionLikeGameCenterBubble];
    } else if (ges.state == UIGestureRecognizerStateChanged) {
        self.frontView.center = dragPoint;
        if (r1 < 6) {
            fillColorForCute = [UIColor clearColor];
            backView.hidden = YES;
            [shapeLayer removeFromSuperlayer];
        }
        [self drawRect];
    } else if (ges.state == UIGestureRecognizerStateEnded || ges.state ==UIGestureRecognizerStateCancelled || ges.state == UIGestureRecognizerStateFailed) {
        backView.hidden = YES;
        fillColorForCute = [UIColor clearColor];
        [shapeLayer removeFromSuperlayer];
        [UIView animateWithDuration:0.5 delay:0.0 usingSpringWithDamping:0.4 initialSpringVelocity:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
            self.frontView.center = oldBackViewCenter;
        } completion:^(BOOL finished) {
            if (finished) {
                [self addAniamtionLikeGameCenterBubble];
            }
        }];
    }
}
//---- 类似GameCenter的气泡晃动动画 ------

-(void)addAniamtionLikeGameCenterBubble {
    
    CAKeyframeAnimation *pathAnimation = [CAKeyframeAnimation animationWithKeyPath:@"position"];
    pathAnimation.calculationMode = kCAAnimationPaced;
    pathAnimation.fillMode = kCAFillModeForwards;
    pathAnimation.removedOnCompletion = NO;
    pathAnimation.repeatCount = INFINITY;
    pathAnimation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear];
    pathAnimation.duration = 5.0;
    
    CGMutablePathRef curvedPath = CGPathCreateMutable();
    CGRect circleContainer = CGRectInset(self.frontView.frame, self.frontView.bounds.size.width / 2 - 3, self.frontView.bounds.size.width / 2 - 3);
    CGPathAddEllipseInRect(curvedPath, NULL, circleContainer);
    
    pathAnimation.path = curvedPath;
    CGPathRelease(curvedPath);
    [self.frontView.layer addAnimation:pathAnimation forKey:@"myCircleAnimation"];
    
    CAKeyframeAnimation *scaleX = [CAKeyframeAnimation animationWithKeyPath:@"transform.scale.x"];
    scaleX.duration = 1;
    scaleX.values = @[@1.0, @1.1, @1.0];
    scaleX.keyTimes = @[@0.0, @0.5, @1.0];
    scaleX.repeatCount = INFINITY;
    scaleX.autoreverses = YES;
    
    scaleX.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
    [self.frontView.layer addAnimation:scaleX forKey:@"scaleXAnimation"];
    
    
    CAKeyframeAnimation *scaleY = [CAKeyframeAnimation animationWithKeyPath:@"transform.scale.y"];
    scaleY.duration = 1.5;
    scaleY.values = @[@1.0, @1.1, @1.0];
    scaleY.keyTimes = @[@0.0, @0.5, @1.0];
    scaleY.repeatCount = INFINITY;
    scaleY.autoreverses = YES;
    scaleY.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
    [self.frontView.layer addAnimation:scaleY forKey:@"scaleYAnimation"];
}

-(void)RemoveAniamtionLikeGameCenterBubble {
    [self.frontView.layer removeAllAnimations];
}

@end

File: AGTIA_Swift/QQCuteView/QQCuteView/QQCuteView.swift
//
//  QQCuteView.swift
//  QQCuteView
//
//  Created by fernando on 2017/2/6.
//  Copyright © 2017年 fernando. All rights reserved.
//

import UIKit

struct BubbleOptions {
    var text: String = ""
    var bubbleWidth: CGFloat = 0.0
    var viscosity: CGFloat = 0.0
    var bubbleColor: UIColor = UIColor.white
}

class QQCuteView: UIView {

    var frontView: UIView?
    var bubbleOptions: BubbleOptions {
        didSet {
            bubbleLabel.text = bubbleOptions.text
        }
    }
    
    private var bubbleLabel: UILabel!
    private var containerView: UIView!

    private var cutePath: UIBezierPath!
    private var fillColorForCute: UIColor!
    private var animator: UIDynamicAnimator!
    private var snap: UISnapBehavior!
    private var backView: UIView!
    private var shapeLayer: CAShapeLayer!
    
    private var r1: CGFloat = 0.0
    private var r2: CGFloat = 0.0
    private var x1: CGFloat = 0.0
    private var y1: CGFloat = 0.0
    private var x2: CGFloat = 0.0
    private var y2: CGFloat = 0.0
    private var centerDistance: CGFloat = 0.0
    private var cosDigree: CGFloat = 0.0
    private var sinDigree: CGFloat = 0.0
    
    private var pointA = CGPoint.zero
    private var pointB = CGPoint.zero
    private var pointC = CGPoint.zero
    private var pointD = CGPoint.zero
    private var pointO = CGPoint.zero
    private var pointP = CGPoint.zero
    
    private var initialPoint: CGPoint = CGPoint.zero
    private var oldBackViewFrame: CGRect = CGRect.zero
    private var oldBackViewCenter: CGPoint = CGPoint.zero

    init(point: CGPoint, superView: UIView, options: BubbleOptions) {
        bubbleOptions = options
        super.init(frame: CGRect(x: point.x, y: point.y, width: options.bubbleWidth, height: options.bubbleWidth))
        initialPoint = point
        containerView = superView
        containerView.addSubview(self)
        setUp()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    fileprivate func drawRect() {
        
        guard frontView != nil else {
            return
        }
        
        x1 = backView.center.x
        y1 = backView.center.y
        x2 = (frontView?.center.x)!
        y2 = (frontView?.center.y)!
        let deltaXSquare = (x2 - x1)*(x2 - x1)
        let deltaYSquare = (y2 - y1)*(y2 - y1)
        centerDistance = sqrt(deltaXSquare + deltaYSquare)
        
        if centerDistance == 0 {
            cosDigree = 1
            sinDigree = 0
        } else {
            cosDigree = (y2 - y1)/centerDistance
            sinDigree = (x2 - x1)/centerDistance
        }
        
        r1 = oldBackViewFrame.size.width/2 - centerDistance/bubbleOptions.viscosity
        pointA = CGPoint(x: x1-r1*cosDigree, y: y1+r1*sinDigree) // A
        pointB = CGPoint(x: x1+r1*cosDigree, y: y1-r1*sinDigree) // B
        pointD = CGPoint(x: x2-r2*cosDigree, y: y2+r2*sinDigree) // D
        pointC = CGPoint(x: x2+r2*cosDigree, y: y2-r2*sinDigree) // C
        pointO = CGPoint(x: pointA.x + (centerDistance / 2)*sinDigree, y: pointA.y + (centerDistance / 2)*cosDigree)
        pointP = CGPoint(x: pointB.x + (centerDistance / 2)*sinDigree, y: pointB.y + (centerDistance / 2)*cosDigree)
        
        backView.center = oldBackViewCenter;
        backView.bounds = CGRect(x: 0, y: 0, width: r1*2, height: r1*2);
        backView.layer.cornerRadius = r1;
        
        cutePath = UIBezierPath()
        cutePath.move(to: pointA)
        cutePath.addQuadCurve(to: pointD, controlPoint: pointO)
        cutePath.addLine(to: pointC)
        cutePath.addQuadCurve(to: pointB, controlPoint: pointP)
        cutePath.move(to: pointA)
        
        if !backView.isHidden {
            shapeLayer.path = cutePath.cgPath
            shapeLayer.fillColor = fillColorForCute.cgColor
            containerView.layer.insertSublayer(shapeLayer, below: frontView?.layer)
        }
    }
    
    fileprivate func setUp() {
        shapeLayer = CAShapeLayer()
        backgroundColor = UIColor.clear
        frontView = UIView(frame: CGRect(x: initialPoint.x, y: initialPoint.y, width: bubbleOptions.bubbleWidth, height: bubbleOptions.bubbleWidth))
        guard frontView != nil else {
            print("frontView is nil")
            return
        }

        r2 = (frontView?.bounds.size.width)! / 2.0
        frontView?.layer.cornerRadius = r2
        frontView?.backgroundColor = bubbleOptions.bubbleColor
        
        backView = UIView(frame: (frontView?.frame)!)
        r1 = backView.bounds.size.width / 2
        backView.layer.cornerRadius = r1
        backView.backgroundColor = bubbleOptions.bubbleColor

        bubbleLabel = UILabel()
        bubbleLabel.frame = CGRect(x: 0, y: 0, width: (frontView?.bounds.width)!, height: (frontView?.bounds.height)!)
        bubbleLabel.textColor = UIColor.white
        bubbleLabel.textAlignment = .center
        bubbleLabel.text = bubbleOptions.text

        frontView?.insertSubview(bubbleLabel, at: 0)
        containerView.addSubview(backView)
        containerView.addSubview((frontView)!)

        x1 = backView.center.x
        y1 = backView.center.y
        x2 = (frontView?.center.x)!
        y2 = (frontView?.center.y)!

        pointA = CGPoint(x: x1-r1, y: y1)   // A
        pointB = CGPoint(x: x1+r1, y: y1)  // B
        pointD = CGPoint(x: x2-r2, y: y2)  // D
        pointC = CGPoint(x: x2+r2, y: y2)  // C
        pointO = CGPoint(x: x1-r1, y: y1)   // O
        pointP = CGPoint(x: x2+r2, y: y2)  // P

        oldBackViewFrame = backView.frame
        oldBackViewCenter = backView.center
        
        backView.isHidden = true //为了看到frontView的气泡晃动效果，需要暂时隐藏backView
        addAniamtionLikeGameCenterBubble()
        
        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(handlePan(ges:)))
        frontView?.addGestureRecognizer(panGesture);
    }
    
    @objc private func handlePan(ges: UIPanGestureRecognizer) {
        let dragPoint = ges.location(in: containerView)
//        if ges.state == .began {
//            // 不给r1赋初始值的话，如果第一次拖动使得r1少于6，第二次拖动就直接隐藏绘制路径了
//            r1 = oldBackViewFrame.width / 2
//            backView.isHidden = false
//            fillColorForCute = bubbleOptions.bubbleColor
//            removeAniamtionLikeGameCenterBubble()
//        } else if ges.state == .changed {
//            frontView?.center = dragPoint
//            if r1 <= 6 {
//                fillColorForCute = UIColor.clear
//                backView.isHidden = true
//                shapeLayer.removeFromSuperlayer()
//            }
//            drawRect()
//        } else if ges.state == .cancelled || ges.state == .failed || ges.state == .ended {
//            backView.isHidden = true
//            fillColorForCute = UIColor.clear
//            shapeLayer.removeFromSuperlayer()
//            
//            UIView.animate(withDuration: 0.5, delay: 0.0, usingSpringWithDamping: 0.4, initialSpringVelocity: 0.0, options: .curveEaseInOut, animations: {
//                self.frontView?.center = self.oldBackViewCenter
//            }, completion: { (finish) in
//                self.addAniamtionLikeGameCenterBubble()
//            })
//        }
        switch ges.state {
        case .began:
            // 不给r1赋初始值的话，如果第一次拖动使得r1少于6，第二次拖动就直接隐藏绘制路径了
            r1 = oldBackViewFrame.width / 2
            backView.isHidden = false
            fillColorForCute = bubbleOptions.bubbleColor
            removeAniamtionLikeGameCenterBubble()
            break
        case .changed:
            frontView?.center = dragPoint
            if r1 <= 6 {
                fillColorForCute = UIColor.clear
                backView.isHidden = true
                shapeLayer.removeFromSuperlayer()
            }
            drawRect()
            break
            
        default:
            backView.isHidden = true
            fillColorForCute = UIColor.clear
            shapeLayer.removeFromSuperlayer()
            
            UIView.animate(withDuration: 0.5, delay: 0.0, usingSpringWithDamping: 0.4, initialSpringVelocity: 0.0, options: .curveEaseInOut, animations: { 
                self.frontView?.center = self.oldBackViewCenter
            }, completion: { (finish) in
                self.addAniamtionLikeGameCenterBubble()
            })
            
            break
        }
    }
}

extension QQCuteView {
    fileprivate func addAniamtionLikeGameCenterBubble() {
        
        let pathAnimation = CAKeyframeAnimation(keyPath: "position")
        pathAnimation.calculationMode = kCAAnimationPaced
        
        pathAnimation.fillMode = kCAFillModeForwards
        pathAnimation.isRemovedOnCompletion = false
        pathAnimation.repeatCount = Float.infinity
        pathAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionLinear)
        pathAnimation.duration = 5.0
        
        let curvedPath = CGMutablePath()
        guard frontView != nil else {
            return
        }
        
        let circleContainer = frontView?.frame.insetBy(dx: (frontView?.bounds.width)! - 3, dy: (frontView?.bounds.height)! - 3)
        curvedPath.addEllipse(in: circleContainer!)
        
        pathAnimation.path = curvedPath
        frontView?.layer.add(pathAnimation, forKey: "circleAnimation")
        
        let scaleX = CAKeyframeAnimation(keyPath: "transform.scale.x")
        scaleX.duration = 1.0
        scaleX.values = [NSNumber(value: 1.0), NSNumber(value: 1.1), NSNumber(value: 1.0)]
        scaleX.keyTimes = [NSNumber(value: 0.0), NSNumber(value: 0.5), NSNumber(value: 1.0)]
        scaleX.repeatCount = Float.infinity
        scaleX.autoreverses = true
        
        scaleX.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        frontView?.layer.add(scaleX, forKey: "scalXAnimation")

        let scaleY = CAKeyframeAnimation(keyPath: "transform.scale.y")
        scaleY.duration = 1.5
        scaleY.values = [NSNumber(value: 1.0),NSNumber(value: 1.1),NSNumber(value: 1.0)]
        scaleY.keyTimes = [NSNumber(value: 0.0), NSNumber(value: 0.5), NSNumber(value: 1.0)]
        scaleY.repeatCount = Float.infinity
        scaleY.autoreverses = true
        scaleY.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        frontView?.layer.add(scaleY, forKey: "scaleYAnimation")
    }
    
    fileprivate func removeAniamtionLikeGameCenterBubble() {
        guard frontView != nil else {
            return
        }
        frontView?.layer.removeAllAnimations()
    }
}
